src/ranges.jl

```diff
@@ -0,0 +1,139 @@
+
+"""
+known_first(::Type{T})
+
+If `first` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_first(typeof(1:4)))
+@test isone(known_first(typeof(Base.OneTo(4))))
+"""
+known_first(x) = known_first(typeof(x))
+known_first(::Type{T}) where {T} = nothing
+known_first(::Type{Base.OneTo{T}}) where {T} = one(T)
+
+"""
+known_last(::Type{T})
+
+If `last` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_last(typeof(1:4)))
+using StaticArrays
+@test known_last(typeof(SOneTo(4))) == 4
+"""
+known_last(x) = known_last(typeof(x))
+known_last(::Type{T}) where {T} = nothing
+
+"""
+known_step(::Type{T})
+
+If `step` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_step(typeof(1:0.2:4)))
+@test isone(known_step(typeof(1:4)))
+"""
+known_step(x) = known_step(typeof(x))
+known_step(::Type{T}) where {T} = nothing
+known_step(::Type{<:AbstractUnitRange{T}}) where {T} = one(T)
+
+_eltype(::Type{T}) where {T} = T
+_eltype(::Type{Val{V}}) where {V} = typeof(V)
+
+struct OptionallyStaticRange{T<:Integer,F,S,L} <: OrdinalRange{T,T}
+  start::F
+  step::S
+  stop::L
+
+  function OptionallyStaticRange(start::F, step::S, stop::L) where {F,S,L}
+      T = promote_type(_eltype(F), _eltype(S), eltype(L))
+      return new{T,F,S,L}(start, step, stop)
+  end
+end
+
+Base.first(r::OptionallyStaticRange{<:Any,Val{F}}) where {F} = F
+Base.first(r::OptionallyStaticRange{<:Any,<:Any}) = getfield(r, :start)
```

Comment:
I'm curious, what is the reason to prefer `getfield(r, :start)` over `r.start`?
I've always used the latter in my own code.

Reply:
Sorry, it's just a habit from working with types that have overloaded the `getproperty` method for metadata like tables. I can change this, as it's probably easier to read `r.start`.